ext/kv/lib.rs

```diff
@@ -0,0 +1,358 @@
+// Copyright 2018-2023 the Deno authors. All rights reserved. MIT license.
+
+pub mod codec;
+mod interface;
+pub mod sqlite;
+
+use std::borrow::Cow;
+use std::cell::RefCell;
+use std::num::NonZeroU32;
+use std::rc::Rc;
+
+use codec::decode_key;
+use codec::encode_key;
+use deno_core::anyhow::Context;
+use deno_core::error::AnyError;
+use deno_core::include_js_files;
+use deno_core::op;
+use deno_core::serde_v8::BigInt;
+use deno_core::ByteString;
+use deno_core::Extension;
+use deno_core::ExtensionBuilder;
+use deno_core::OpState;
+use deno_core::Resource;
+use deno_core::ResourceId;
+use deno_core::ZeroCopyBuf;
+use serde::Deserialize;
+use serde::Serialize;
+
+pub use crate::interface::*;
+
+fn build_ops<DBH: DatabaseHandler + 'static>(
+  handler: DBH,
+  builder: &mut ExtensionBuilder,
+) -> &mut ExtensionBuilder {
+  let handler = Rc::new(handler);
+  builder
+    .ops(vec![
+      op_kv_database_open::decl::<DBH>(),
+      op_kv_snapshot_read::decl::<DBH::DB>(),
+      op_kv_atomic_write::decl::<DBH::DB>(),
+      op_kv_encode_key::decl(),
+      op_kv_decode_key::decl(),
+    ])
+    .state(move |state| {
+      state.put(handler.clone());
+    })
+}
+
+pub fn init_ops<DBH: DatabaseHandler + 'static>(handler: DBH) -> Extension {
+  build_ops(handler, &mut Extension::builder(env!("CARGO_PKG_NAME"))).build()
+}
+
+pub fn init_ops_and_esm<DBH: DatabaseHandler + 'static>(
+  handler: DBH,
+) -> Extension {
+  build_ops(
+    handler,
+    Extension::builder(env!("CARGO_PKG_NAME"))
+      .esm(include_js_files!("01_db.js",)),
+  )
+  .build()
+}
+
+struct DatabaseResource<DB: Database + 'static> {
+  db: Rc<DB>,
+}
+
+impl<DB: Database + 'static> Resource for DatabaseResource<DB> {
+  fn name(&self) -> Cow<str> {
+    "database".into()
+  }
+}
+
+#[op]
+async fn op_kv_database_open<DBH>(
+  state: Rc<RefCell<OpState>>,
+  path: Option<String>,
+) -> Result<ResourceId, AnyError>
+where
+  DBH: DatabaseHandler + 'static,
+{
+  let handler = {
+    let state = state.borrow();
+    state.borrow::<Rc<DBH>>().clone()
+  };
+  let db = handler.open(path).await?;
```

Comment:
I think we need to check --read and --write permissions for this file here? Also, we need unstable checks. You can find some examples in the code by searching for "check_unstable"

Reply:
Agreed, we should also have at least one test case that verifies that permissions are checked

Reply:
The default database is treated like localStorage - it can be stored in the DENO_DIR